core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java

```diff
@@ -123,189 +151,344 @@ public void unableToMonitorLockNode(Exception e) {
+                zooKeeper.getZooKeeper().removeWatches(nodeToWatch, this, WatcherType.Any, true);
+                determineLockOwnership(createdEphemeralNode, lw);
+              }
+            } catch (KeeperException | InterruptedException e) {
+              lw.failedToAcquireLock(new Exception("Failed to renew watch on other master node"));
             }
-          } catch (KeeperException | InterruptedException e) {
-            lw.failedToAcquireLock(new Exception("Failed to renew watch on other master node"));
           }
         }
-      }
 
-    });
+      };
+
+      Stat stat = zooKeeper.getStatus(nodeToWatch, priorNodeWatcher);
+      if (stat == null) {
+        // if stat is null from the zookeeper.exists(path, Watcher) call, then we just
+        // created a Watcher on a node that does not exist. Delete the watcher we just created.
+        zooKeeper.getZooKeeper().removeWatches(nodeToWatch, priorNodeWatcher, WatcherType.Any,
+            true);
+        determineLockOwnership(createdEphemeralNode, lw);
+      }
+    }
 
-    if (stat == null)
-      lockAsync(myLock, lw);
   }
 
   private void lostLock(LockLossReason reason) {
     LockWatcher localLw = lockWatcher;
-    lock = null;
+    lockNodeName = null;
     lockWatcher = null;
 
     localLw.lostLock(reason);
   }
 
-  public synchronized void lockAsync(final AsyncLockWatcher lw, byte[] data) {
+  public synchronized void lock(final AccumuloLockWatcher lw, byte[] data) {
 
-    if (lockWatcher != null || lock != null || asyncLock != null) {
+    if (lockWatcher != null || lockNodeName != null || createdNodeName != null) {
       throw new IllegalStateException();
     }
 
     lockWasAcquired = false;
 
     try {
-      final String asyncLockPath = zooKeeper.putEphemeralSequential(path + "/" + LOCK_PREFIX, data);
-      log.trace("Ephemeral node {} created", asyncLockPath);
-      Stat stat = zooKeeper.getStatus(asyncLockPath, new Watcher() {
+      final String lockPathPrefix = path + "/" + vmLockPrefix.toString();
+      // Implement recipe at https://zookeeper.apache.org/doc/current/recipes.html#sc_recipes_Locks
+      // except that instead of the ephemeral lock node being of the form guid-lock- use lock-guid-.
+      // Another deviation from the recipe is that we cleanup any extraneous ephemeral nodes that
+      // were created.
+      final String createPath = zooKeeper.putEphemeralSequential(lockPathPrefix, data);
+      LOG.debug("[{}] Ephemeral node {} created", vmLockPrefix, createPath);
+
+      // It's possible that the call above was retried several times and multiple ephemeral nodes
+      // were created but the client missed the response for some reason. Find the ephemeral nodes
+      // with this ZLOCK_UUID and lowest sequential number.
+      List<String> children = new ArrayList<>();
+      zooKeeper.getChildren(path).forEach(s -> {
+        if (s.startsWith(ZLOCK_PREFIX)) {
+          children.add(s);
+        }
+      });
```

Comment:
This pattern is repeated twice in the code.  Do you think it would make sense to put it in a function and add a stronger check of the nodes than just the prefix (like the pattern `zlock#<uuid>#<seq>`?  Also wondering if something should be logged for child nodes that do not match the pattern, but I am not sure if that is expected.

Reply:
moved to a function and logged invalid format.

Reply:
Regarding the stronger check, I could do that and remove the code for handling invalid formats. I think it depends on how rigid we want the code to be. TBH, I'm not sure whether or not ephemeral nodes might be created at the same location by a different process at the same path.

Reply:
Seems like ignoring/filtering out the unexpected nodes and logging a warning would be good.

Reply:
I merged the methods that gets the children and sorts them, into one [here](https://github.com/apache/accumulo/pull/1866/files#diff-d53a0e8d1a6246f0196696217ef667f8730e51b93ffdea614d380315094e2387R176). Unexpected nodes are not added to the child set and there is a log message as to why.